subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -198,4 +204,25 @@ public Configuration getCompileDependencies() {
             return convention.getProject().getConfigurations().getByName(JavaPlugin.COMPILE_CONFIGURATION_NAME);
         }
     }
+
+    private class SourceSetOutputCleanUpRegistrationListener implements ProjectEvaluationListener {
```

Comment:
If registration would support somewhat the same things as `org.gradle.api.Project#files(java.lang.Object...)` - i.e. some lazy evaluation - we could register the outputs right when we apply the plugin and evaluate them whenever we want to delete them. If we don't do that then we have the problem to decide when we want to evaluate the outputs. We would only evaluate all those paths when we really need to clean. WDYT?